### PR TITLE
Default project when cancelling project loading

### PR DESCRIPTION
**Pull request by zonkmachine**
If a user isn't aware that the loading of a project has been
cancelled, it may be in an incomplete state. Saving such a project
will overwrite the original file and result in data loss. This is
solved by loading the default project on cancelling project loading.

Fixes https://github.com/LMMS/lmms/issues/3636

### CODE REVIEW

File: src/core/Song.cpp [hunk 1]
```diff
@@ -1150,6 +1151,12 @@ void Song::loadProject( const QString & fileName )
 	m_modified = false;
 	m_loadOnLaunch = false;
 
+	if( isCancelled() )
+	{
+		m_isCancelled = false;
+		createNewProject();
```
Comment (zonkmachine) [inline src/core/Song.cpp]: This should be done earlier. At least before [the error summary](https://github.com/LMMS/lmms/pull/3941/files#diff-bc7aaaf624969ba372535c9eab5ed490R1137).

**Comment by PhysSong** [comment]
Seems to work fine. However, I want to test this more for some reason. After it, I'll either approve it or request some changes.

**Comment by zonkmachine** [comment]
I've tested to not adding cancelled programs to the recent files list but found that rather unintuitive actually. The brain seem to add it in there for my inner vision anyway and when I reach for the recent project list I'm 'off by one'.

**Comment by PhysSong** [comment]
@zonkmachine Do you have a backtrace? Can you provide some more details?

**Comment by zonkmachine** [comment]
```
(gdb) bt full
#0  0x0000000000523362 in InstrumentPlayHandle::play (this=0x1c93c5c0, _working_buffer=0x1000cf0020)
    at /home/zonkmachine/builds/lmms/lmms/include/InstrumentPlayHandle.h:46
        nphv = {{p = {static shared_null = {ref = {_q_value = 12591}, alloc = 0, begin = 0, end = 0, sharable = 1, array = {0x0}}, d = 0x7fffb1e47c30}, 
            d = 0x7fffb1e47c30}}
        nphsLeft = false
#1  0x0000000000544abf in PlayHandle::doProcessing (this=0x1c93c5c0) at /home/zonkmachine/builds/lmms/lmms/src/core/PlayHandle.cpp:58
No locals.
#2  0x0000000000523221 in ThreadableJob::process (this=0x1c93c5c0) at /home/zonkmachine/builds/lmms/lmms/include/ThreadableJob.h:74
No locals.
#3  0x00000000005393b3 in MixerWorkerThread::JobQueue::run (this=0x9634e0 <MixerWorkerThread::globalJobQueue>)
    at /home/zonkmachine/builds/lmms/lmms/src/core/MixerWorkerThread.cpp:74
        job = 0x1c93c5c0
        i = 0
        processedJob = false
#4  0x000000000053960d in MixerWorkerThread::startAndWaitForJobs () at /home/zonkmachine/builds/lmms/lmms/src/core/MixerWorkerThread.cpp:147
No locals.
#5  0x00000000005340c7 in Mixer::renderNextBuffer (this=0xc44190) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:440
        playModeSupportsMetronome = false
        it_rem = {i = 0x7fffa80013e8}
        fxMixer = 0xc2e3c0
        last_metro_pos = {<MidiTime> = {m_ticks = -1, static s_ticksPerTact = 192}, m_timeLine = 0x0, m_currentFrame = 0}
        song = 0xc4bf80
        currentPlayMode = Song::Mode_None
        p = {<MidiTime> = {m_ticks = 0, static s_ticksPerTact = 192}, m_timeLine = 0x0, m_currentFrame = 0}
#6  0x0000000000535c2b in Mixer::fifoWriter::run (this=0x10e50f0) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:1103
        buffer = 0x7fffa80031f0
        b = 0xc8c440
        frames = 256
#7  0x00007ffff685a32f in QThreadPrivate::start (arg=0x10e50f0) at thread/qthread_unix.cpp:349
        __clframe = {__cancel_routine = 0x7ffff6859300 <QThreadPrivate::finish(void*)>, __cancel_arg = 0x10e50f0, __do_it = 1, 
          __cancel_type = <optimized out>}
        thr = 0x10e50f0
        data = <optimized out>
        objectName = {static null = {<No data fields>}, static shared_null = {ref = {_q_value = 18604}, alloc = 0, size = 0, 
            data = 0x7ffff6cc293a <QString::shared_null+26>, clean = 0, simpletext = 0, righttoleft = 0, asciiCache = 0, capacity = 0, reserved = 0, 
            array = {0}}, static shared_empty = {ref = {_q_value = 180}, alloc = 0, size = 0, data = 0x7ffff6cc291a <QString::shared_empty+26>, clean = 0, 
            simpletext = 0, righttoleft = 0, asciiCache = 0, capacity = 0, reserved = 0, array = {0}}, d = 0x7ffff6cc2920 <QString::shared_null>, 
          static codecForCStrings = 0x0}
#8  0x00007ffff7bc4184 in start_thread (arg=0x7fffb1e48700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffb1e48700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736177932032, 2699179435038421571, 1, 0, 140736177932736, 140736177932032, -2699333285049191869, 
                -2699197564588445117}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#9  0x00007ffff3f8affd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

```

**Comment by zonkmachine** [comment]
` QThreadPrivate::start (`

So it kicks off a new thread somewhere.

**Comment by zonkmachine** [comment]
What I was doing is cancelling the project at the very end of loading unfa/spoken. I've seen this more than once and only after hitting the 90% mark.

**Comment by PhysSong** [comment]
I got it. I think you were loading a project with single-streamed instrument(s) and canceled it. I have a simple way to fix that, related to mixer. May I just push that to your branch tomorrow?

**Comment by zonkmachine** [comment]
> May I just push that to your branch tomorrow?

You may.

File: src/core/Song.cpp [hunk 2]
```diff
@@ -1133,6 +1134,13 @@ void Song::loadProject( const QString & fileName )
 
 	emit projectLoaded();
 
+	if( isCancelled() )
```
Comment (PhysSong) [inline src/core/Song.cpp]: I think this should be right after the `while` loop. But then LMMS will crash if you cancel loading a project which has single-streamed instruments. However, I have a fix for this as I mentioned before, and I'm going to push this today.

Anyway it works nice for projects without such instruments. :+1:

**Comment by PhysSong** [comment]
FYI, it fixes #3670 too.

**Comment by PhysSong** [comment]
Instead of direct push, I'll provide a patch file for that. I was a little bit busy...
> I've tested to not adding cancelled programs to the recent files list but found that rather unintuitive actually.

What do you exactly mean there?

**Comment by zonkmachine** [comment]
>> I've tested to not adding cancelled programs to the recent files list but found that rather unintuitive actually.
> 
> What do you exactly mean there?

Not very important. If I break loading before [this line](https://github.com/LMMS/lmms/blob/7ed9bea9fbded61b4ba2d32113e442c38f84a9b9/src/core/Song.cpp#L1130), or move it to after the break, the project will not be bumped up to the top of the recent projects list. I tried this and was surprised at how unintuitive it felt. I'm pretty sure I would get used to it quickly but as a first impression it was a bit odd.

**Comment by zonkmachine** [comment]
> Instead of direct push, I'll provide a patch file for that.

Aight, works fine for me!

**Comment by PhysSong** [comment]
> the project will not be bumped up to the top of the recent projects list.

Looks minor, but a little bit tricky... Users may cancel project loading because of several reasons. For some situation(ex. clicked wrong file), that behavior is appropriate, but current one may be better for other situations.

**Comment by tresf** [comment]
> Not very important. If I break loading before this line, or move it to after the break, the project will not be bumped up to the top of the recent projects list. I tried this and was surprised at how unintuitive it felt. I'm pretty sure I would get used to it quickly but as a first impression it was a bit odd.

As a baseline comparison I just tested this with Microsoft Excel (it takes a while to load as well) and the behavior is identical to what @zonkmachine describes feels normal.  i.e. Even if you cancel the loading, it still shows at the top of recent list.

**Comment by PhysSong** [comment]
Sorry to say that, but I lost my works by accident. Adding a function in `Mixer` class will be useful.
```c++
void Mixer::clearNewPlayHandles()
{
	requestChangeInModel();
	for( LocklessListElement * e = m_newPlayHandles.popList(); e; )
	{
		LocklessListElement * next = e->next;
		m_newPlayHandles.free( e );
		e = next;
	}
	doneChangeInModel();
}
```
To fix the crash, call this function when project loading is canceled.

**Comment by zonkmachine** [comment]
> To fix the crash, call this function when project loading is canceled.

OK, thanks! https://github.com/LMMS/lmms/pull/3941/commits/14182a310b337a5f68aa879478b8ed826de89b8b
I'm calling `Mixer::clearNewPlayHandles()` from Song.h and add Mixer.h as a dependency. No more crashes reproducible. :)

**Comment by zonkmachine** [comment]
> For some situation(ex. clicked wrong file), that behavior is appropriate,

This was the reason I tried not adding it to the list. The current behaviour is to add a project to the recent list even if it's cancelled. I think it's best to leave it as it is for now though I have no real opinion on how this should best be handled. If someone feels strongly about it then I think it's better to look into it for 1.3 or later.

Merging.

**Pull request merged**